### PR TITLE
[Impeller] switch Rect fields to LTRB implementation

### PR DESCRIPTION
**Pull request by flar**
Switches the Rect class implementation to an LTRB (Left, Top, Right, Bottom) set of fields for efficiency in performing rectangle operations. Additionally, the methods have been somewhat hardened to the following issues:

- NaN values should always act as if the rect is empty
  - Helps with https://github.com/flutter/flutter/issues/132770
- Saturated math methods are added to avoid overflow for integer rects
- Normalized treatment of empty rectangles in rect/rect operations
  - empty.Union(anything) == anything
  - empty.Intersection(anything) == empty
  - empty.IntersectsWith(anything) == false
  - empty.Contains(anything) == false
  - non-empty.Contains(empty) == true
- MakeMaximum now returns finite rectangles which should produce non-nan values from all getters
- A lot more testing of all of these behaviors within the unit tests

In addition, the new rectangle is closer to what DisplayList and the Skia backend have been using all along so we might be able to switch to using Impeller geometry inside DisplayList to reduce overhead for both backends.

### CODE REVIEW

File: impeller/geometry/rect.h [hunk 1]
```diff
@@ -69,76 +166,114 @@ struct TRect {
...
+    return Scale(scale.x, scale.y);
   }
 
   [[nodiscard]] constexpr TRect Scale(TSize<T> scale) const {
-    return Scale(TPoint<T>(scale));
+    return Scale(scale.width, scale.height);
   }
 
+  /// @brief  Returns true iff the provided point |p| is inside the
+  ///         helf-open interior of this rectangle.
```
Comment (jonahwilliams) [inline impeller/geometry/rect.h:206]:
```suggestion
  ///         half-open interior of this rectangle.
```

File: impeller/geometry/rect.h [hunk 2]
```diff
@@ -69,76 +166,114 @@ struct TRect {
...
+  ///         An empty rectangle can contain no other rectangle.
+  ///
+  ///         An empty rectangle is, however, contained within any
+  ///         other non-empy rectangle as the set of points it contains
+  ///         is an empty set and so there are no points to fail the
+  ///         containment criteria.
   [[nodiscard]] constexpr bool Contains(const TRect& o) const {
-    return o.GetLeft() >= GetLeft() && o.GetTop() >= GetTop() &&
-           o.GetRight() <= GetRight() && o.GetBottom() <= GetBottom();
+    // Is this->IsEmpty test needed here? Write tests and then evaluate if
```
Comment (jonahwilliams) [inline impeller/geometry/rect.h:236]: Add a TODO for this?

File: impeller/geometry/rect_unittests.cc [hunk 3]
```diff
@@ -204,10 +962,10 @@ TEST(RectTest, IRectScale) {
 
   test2(IRect::MakeLTRB(10, 15, 100, 150), 2, 3);
   test2(IRect::MakeLTRB(10, 15, 100, 150), 3, 2);
-  test2(IRect::MakeLTRB(10, 15, -100, 150), 2, 3);
-  test2(IRect::MakeLTRB(10, 15, 100, -150), 2, 3);
-  test2(IRect::MakeLTRB(10, 15, 100, 150), -2, 3);
-  test2(IRect::MakeLTRB(10, 15, 100, 150), 2, -3);
+  // test2(IRect::MakeLTRB(10, 15, -100, 150), 2, 3);
```
Comment (jonahwilliams) [inline impeller/geometry/rect_unittests.cc:965]: Comment/TODO?

**Review by jonahwilliams — COMMENTED** [review]
I've only taken a quick look through the good, but overall it looks great.

LMK if there is anything I can do to help get this landed.

**Comment by flar** [comment]
> I've only taken a quick look through the good, but overall it looks great.
> 
> LMK if there is anything I can do to help get this landed.

It looks like the tests are failing on Intel platforms due to overflow casts from the ceil method returning the wrong answer. It looks like I'll need to add some saturated::Floor and Ceil methods to work around this.

Also, only on the mac_host_engine, not on the Intel Linux builds, it looks like the saturated Add and Sub are failing to saturate correctly.

I'm looking into the failure that happens on Intel Linux with a cloud VM, but I'll have to find access to an Intel Mac somewhere to figure out why the Add and Sub aren't saturating.

File: impeller/geometry/rect.h (hunk 2 above)
Comment (flar) [reply impeller/geometry/rect.h:236]: Actually, it ended up being needed after I adjusted the semantics so it's just obsolete now.

File: impeller/geometry/rect.h (hunk 1 above)
Comment (flar) [reply impeller/geometry/rect.h]: Done

File: impeller/geometry/rect.h (hunk 2 above)
Comment (flar) [reply impeller/geometry/rect.h]: Done

File: impeller/geometry/rect_unittests.cc (hunk 3 above)
Comment (flar) [reply impeller/geometry/rect_unittests.cc]: Reactivated the tests.

**Comment by flar** [comment]
I've implemented saturated casts which should help with most of the previous failures, but I still have to track down the failures in saturated Add/Sub on Intel Macs.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by jason-simmons — APPROVED** [review]

**Pull request merged**